components/clp-py-utils/clp_py_utils/s3_utils.py

```diff
@@ -1,18 +1,177 @@
-import re
+import os
 from pathlib import Path
-from typing import List, Tuple
+from typing import Dict, List, Literal, Optional, Tuple
 
 import boto3
 from botocore.config import Config
 from job_orchestration.scheduler.job_config import S3InputConfig
 
-from clp_py_utils.clp_config import S3Config
+from clp_py_utils.clp_config import CLPConfig, S3Config, S3Credentials, StorageType
 from clp_py_utils.compression import FileMetadata
 
 # Constants
 AWS_ENDPOINT = "amazonaws.com"
 
 
+def _get_session_credentials(aws_profile: Optional[str] = None) -> Optional[S3Credentials]:
+    """
+    Generates AWS credentials created by boto3 when starting a session with given profile.
+    :param aws_profile: Name of profile configured in ~/.aws directory
+    :return: An S3Credentials object with access key pair and session token if applicable.
+    """
+    aws_session = None
+    if aws_profile is not None:
+        aws_session = boto3.Session(profile_name=aws_profile)
+    else:
+        aws_session = boto3.Session()
+    credentials = aws_session.get_credentials()
+    if credentials is None:
+        return None
+
+    return S3Credentials(
+        access_key_id=credentials.access_key,
+        secret_access_key=credentials.secret_key,
+        session_token=credentials.token,
+    )
+
+
+def get_credential_env_vars(config: S3Config) -> Dict[str, str]:
+    """
+    Generates AWS credential environment variables for tasks.
+    :param config: S3Config or S3InputConfig from which to retrieve credentials.
+    :return: A [str, str] Dict which access key pair and session token if applicable.
+    :raise: ValueError if auth type is not supported
+    """
+    env_vars: Dict[str, str] = None
+    auth = config.aws_authentication
+
+    if "profile" == auth.type:
+        aws_credentials: S3Credentials = _get_session_credentials(auth.profile)
+        if aws_credentials is None:
+            raise ValueError(f"Failed to authenticate with profile {auth.profile}")
+
+        env_vars = {
+            "AWS_ACCESS_KEY_ID": aws_credentials.access_key_id,
+            "AWS_SECRET_ACCESS_KEY": aws_credentials.secret_access_key,
+        }
+        aws_session_token = aws_credentials.session_token
+        if aws_session_token is not None:
+            env_vars["AWS_SESSION_TOKEN"] = aws_session_token
```

Comment:
this part of code is duplicated in 3 clauses. Can we reorganize the code by doing the following:
1. Init credentials: S3Credentials;
2. Handle "env_vars" == auth.type: with early return.
3. For credentials, ec2 and profile, update the "credentials" in their own if-statement.
4. use the common logic to convert it into env_vars format.